[1단계 - 지하철 정보 관리 기능] 루카(백경환) 미션 제출합니다.

dooboocookie opened the pull request:
안녕하세요! 제이온!!

이번에 시간이 많이 부족해서 완성도 있는 코드와 테스트를 짜지 못하였습니다.
기능이 돌아가는 데에만 집중해서, service에 비즈니스 로직이 절차지향적으로 몰려 있는 상황입니다.. 😭
그래서 계속 리팩토링하고 도메인도 만들면서, 여러 고민을 해봐야 할 것 같습니다...

지금도 급하게 제출하느라, 질문할 점을 생각하지 못했습니다...
리뷰 잘 부탁드리겠습니다.!!🙇‍♂️🙇‍♂️🙇‍♂️

hyena0608 committed:
docs: 지하철 요구사항 정리

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
test: 지하철역 저장 인수 테스트

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: 역, 구간, 거리 도메인 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: 역 저장, 조회, 삭제 기능 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: 노선 저장, 조회, 삭제 기능 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: 역_노선 저장, 조회, 삭제 기능 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: 구간 저장, 조회, 삭제 기능 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: DB 테이블 DML, 더미 데이터 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: 데이터 변환용 객체 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: 구간 저장 기능 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: 구간 삭제 로직 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: 노선의 역 목록 조회 기능 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: 전체 노선 조회 기능 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: POST /stations 구간 저장 기능 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: GET /lines/{lineName}/stations 노선 조회 기능 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
feat: GET /lines/stations 전체 노선 조회 기능 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
test: 구간, 노선, 역 테스트 추가

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

hyena0608 committed:
chore: 자바 컨벤션

Co-authorized-by: dooboocookie <bkh3030@gmail.com>

dooboocookie committed:
docs: 요구사항 목록 삭제

dooboocookie committed:
docs: 요구사항 수정

- 리팩토링 전 API 정리

dooboocookie committed:
docs: 요구사항 수정

- 리팩토링 전 API 정리

dooboocookie committed:
docs: 도메인 요구사항 추가

dooboocookie committed:
feat: 두 거리의 합이 같은지 비교하는 기능 추가

- 테스트 수정

dooboocookie committed:
refactor: Station의 Sections 정보 재거

dooboocookie committed:
feat: Section 사이의 중간 구간들 추가 기능, 두 Section을 병합하는 기능 추가

- Distance 거리 합치는 기능 추가
- Section의 노선 정보 삭제
- Section의 시작점 여부 삭제

pjy1368 commented on src/main/resources/schema.sql:
```diff
@@ -1,14 +1,33 @@
-create table if not exists STATION
+create table if not exists stations
```
테이블 이름을 복수형으로 변경하신 이유가 있나요?

dooboocookie replied:
테이블 이름을 복수형으로 한 명확한 소신이 있었던 것은 아닙니다.
일단 페어와 정할 때는 `order`같은 예약어를 피하기에 복수형이 더 적합하다 생각하였습니다!

pjy1368 replied:
예약어가 있는 것이 불편하긴하나 관례적으로 테이블도 하나의 단일 개체로 보는 것이 자연스러워서 단수명으로 네이밍을 많이 가져갑니다. sql에서 예약어를 써야한다면 '₩'(백틱)를 사용하여 표기하시면 됩니다.

pjy1368 commented on src/main/resources/schema.sql:
```diff
@@ -1,14 +1,33 @@
+    primary key (id)
+);
+
+create table if not exists sections
+(
+    id              bigint not null auto_increment,
+    distance        int                   not null,
+    is_start        boolean               not null,
+    up_station_id   bigint                not null,
+    down_station_id bigint                not null,
+    line_id         bigint                not null,
+    primary key (id)
```
상행역, 하앵역에 대해 외래키를 설정하지 않으신 이유가 있으실까요?

dooboocookie replied:
외래키를 작성하지 않은 이유는, 테이블 간 결합도를 낮춰 프로젝트를 유동적으로 진행하기 위함이였습니다.

현재는 딱히 외래키를 넣어 불편한 점이나 문제가 됐던 점은 없는 것 같아서 다시 외래키를 설정하였습니다.

pjy1368 replied:
실무에서 외래키를 마이그레이션의 불편함과 데드락의 위험성때문에 거의 설정하지는 않지만, 우리는 연습 단계이므로 외래키 사용을 추천드립니다 ㅎㅎ

pjy1368 commented on src/test/java/subway/integration/StationIntegrationTest.java:
```diff
@@ -1,197 +1,51 @@
-        assertThat(response.statusCode()).isEqualTo(HttpStatus.CREATED.value());
-        assertThat(response.header("Location")).isNotBlank();
-    }
+class StationIntegrationTest extends IntegrationTest {
```
현재는 main의 application.properties의 DB 연결 정보를 읽어서 테스트하고 있는데, 만약 main DB가 H2가 아닌 MySQL 등 다른 DB를 사용하게 된다면 이 E2E 테스트는 어떻게 될까요?

dooboocookie replied:
이 부분은 2단계 요구사항이기도 하고, 제가 코드를 늦게 작성하다 보니 시간상 문제가 생겨서 2단계 때 진행하도록 하겠습니다 😭😭😭😭😭😭😭😭😭😭😭😭😭😭

pjy1368 commented on src/test/java/subway/application/LineServiceTest.java:
```diff
@@ -0,0 +1,119 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+@JdbcTest
+class LineServiceTest {
```
시간이 되신다면 DAO 단위 테스트를 추가하시고, Service에서 DAO를 모킹한 형태의 단위 테스트도 작성해 보시면 좋겠습니다 :)

dooboocookie replied:
넵!! Service, Repository, DAO의 단위 테스트를 모두 작성하고 Service와 Repository는 모킹을 사용하여 진행해보았습니다!!

다만, 모킹으로 테스트를 진행하다보니까, 특정 행위를 하는지에 대한 검증을 많이했는데 잘 진행한건지 모르겠습니다.

테스트에서 너무 많은 설정(?)을 하고 있다는 느낌이 듭니다.

pjy1368 commented on src/main/java/subway/dao/dto/SectionDto.java:
```diff
@@ -0,0 +1,44 @@
+package subway.dao.dto;
+
+public class SectionDto {
+
+    private final Long id;
+    private final Long upStationId;
+    private final Long downStationId;
+    private final Long lineId;
+    private final Integer distance;
+    private final Boolean isStart;
```
Entity 역할을 하는 것 같은데 패키지도 dto로 변경하고 객체 이름도 Entity 혹은 Row로 지어주면 어떨까요?

dooboocookie replied:
넵! Entity와 Domain객체를 분리하여 Dao에서는 Entity를 주고받도록 짰습니다!

pjy1368 reviewed (changes requested):
안녕하세요 루카! 이번 미션의 리뷰어를 맡은 제이온입니다 😊
복잡한 지하철 정보 관리 기능을 구현하기 위해 고생하신 흔적이 느껴지네요 😂
정해진 시간 안에 기능 동작하게끔 잘 해주셨으나, 말씀주신대로 Service의 절차 지향적인 코드를 도메인에 많이 넘겨주도록 수정하시고 다시 리뷰 요청주시면 좋을 것 같아요 ㅎㅎ

dooboocookie committed:
feat: Line 이 Sections을 가지도록 변경 및 기능 추가

- 역(구간) 추가 기능
- 역 삭제 기능
- 역들 조회 기능

dooboocookie committed:
feat: Sections 추가

- List<Section> 일급 컬렉션
- Section들 사이에 역 추가하고, 삭제하는 기능 이동
- Station, Section Empty 객체 상수 등록

dooboocookie committed:
refactor: DAO 와 주고 받는 객체 Entity로 분리 및 DB 설계 수정

dooboocookie committed:
refactor: LineRepository 조회와 변경 업데이트 기능, StationRepository 저장 기능 추가

dooboocookie committed:
refactor: Service 추가 로직 도메인과 레파지토리 활용하도록 변경

dooboocookie committed:
refactor: StationService 삭제 로직 도메인과 레파지토리 활용하도록 변경

- StationController 서비스 메서드 잘 호출하도록 변경

dooboocookie committed:
refactor: LineService에서 도메인과 레포지토리 활용하도록 변경

dooboocookie committed:
feat: 노선에 첫 역 2개 등록하는 기능 추가

dooboocookie committed:
feat: 노선 추가 삭제 기능 추가

dooboocookie committed:
docs: API 설계 수정

dooboocookie committed:
refactor: Station 관련 CRUD 기능 변

- 역 CRUD는 노선관 별개로 등록

dooboocookie committed:
refactor: Line 관련 CRUD 기능 추가

- Subway 검증 로직 추가

dooboocookie committed:
refactor: Section CRD 기능 추가

- 필요없는 메서드 정리

dooboocookie committed:
test: 도메인 기능 테스트

- 도메인 코드 가독성
- 불필요한 메서드 정리
- Subway 검증로직 정리

dooboocookie committed:
test: StationDao 테스트 추가

dooboocookie committed:
test: LineDao 테스트 추가

dooboocookie committed:
test: SectionDao 테스트 추가

dooboocookie committed:
test: LineRepository 테스트 추가

dooboocookie committed:
test: SectionRepositoryTest 테스트 추가

dooboocookie committed:
test: StationService 테스트 추가

dooboocookie committed:
test: LineService 테스트 추가

dooboocookie committed:
test: LineStationService 테스트 추가

dooboocookie committed:
test: 통합 테스트 추가

dooboocookie committed:
chore: 자바 컨벤션

dooboocookie commented:
안녕하세요!! 제이온!!😀😀
백엔드 5기 루카입니다!!

제게 이번 미션이 큰 고비 같은 느낌이 있습니다.
페어와도 많은 이야기를 하였고, 페어와 헤어지고 나서도 `API 설계`, `도메인 설계`, `테이블 설계`를 많이 뒤집었습니다.
그렇게 삽질(?)을 많이 하다보니 피드백 적용도 매우 늦어졌습니다...
늦은만큼 최대한 열심히 작성해보았습니다🥲
이전 PR에 비해 달라진 점이 많아서 리뷰하기 힘드실 것 같다는 걱정이 듭니다.. 미리 죄송합니다... 🙇‍♂️🙇‍♂️

---
# 🧐 중요하게 생각한 부분

## 1. 🆔❓ `테이블의 ID` vs `도메인의 ID` vs `리소스의 ID` 

#### 상황
- 처음에는 `도메인 객체`와 `엔티티` 개념을 분리하지 않고 하나로 사용하려 했습니다
- 그렇다 보니, 도메인 객체의 필드에 따라 DAO에서 영속화를 하거나 조회를 하는 과정에서 JOIN문이 많이 쓰이는 등 DAO가 하는 일 `많아지고 복잡해지는 문제`가 발생하였습니다.
- 그래서 `Respository`라는 계층을 만들어서, `도메인 객체`를 주고 받도록 하고, Repository에서는 `DAO`를 호출하도록 하였습니다.

#### 문제
- `도메인`을 레벨 1 때 콘솔 애플리케이션을 짜듯, ID 없이 작성하였습니다.
- `테이블`에는 `도메인객체와 상응하는 row`가 존재하였습니다.
- 여기서 문제는 도메인 객체를 DB에 영속화할 때 해당 도메인의 아이디 필드가 없어서 어떤 도메인인지 몰랐습니다.
- `name`같은 유니크하다고 가정한 필드로 식별할 수 있었지만 그것은 완벽한 식별자로서 역할은 못한다 생각하였습니다.

#### 현재
- 도메인에도 아이디 필드를 추가하였습니다.
- DB를 완전히 제거하고 생각해도 도메인 내에서도 식별할 수 있는 필드가 필요하다고 생각했습니다
  - (현재 id를 사용하는 도메인로직은 거의 없지만)
- 디비에서 오토 인크리먼트 되어 생성된 id값을 필드로 사용하여 DB가 도메인까지 침범했다는 느낌이 강하게 들지만, 
  - 이는 편의상 그렇게 사용하는 것이라고 생각합니다.
  - 도메인 객체를 영속화하기 전에는 ID가 null인 상태로 사용하는 것이 불안한 문제도 동반한다고 생각합니다.
- 그래서 현재 제가 내린 결론은,
  - 리소스를 식별하기 위해 사용하는 `/{lineId}` 같은 값과
  - 도메인을 식별하기 위해 사용하는 `private final Long id;` 와
  - 테이블의 로우를 식별하기 위해 사용하는 `primary key(id)`는
  - 각 영역에서 식별을 필요로 하기 때문에 존재하는 것이고 현재는 편의상 디비에서 생성된 값을 사용하는 것이라고 판단했습니다.
  - 이에 대한 의견을 여쭙고 싶습니다.

## 2. 🪢Repository의 역할과 Domain끼리의 의존

#### ERD

![image](https://github.com/woowacourse/jwp-subway-path/assets/79090478/1691ad2e-e1aa-4cc2-a680-3211bdf4a0c2)

#### 클래스 다이어그램

![image](https://github.com/woowacourse/jwp-subway-path/assets/79090478/27d22f77-e739-43e5-a108-8eb7a4775dd7)

- 현재 DB의 참조키와 클래스의 의존관계가 약간은 다릅니다.
- `레파지토리`에서 하나의 도메인을 만들기 위해, `여러 DAO`에 접근을 하여 도메인을 `조립`하여 만듭니다.
- 그 과정이 때론 너무 무거워 보입니다.
- 현재 제 레파지토리가 제 역할을 하고 있는지 약간의 의문이 있습니다.
- 문제를 느낀 부분은 2가지 있습니다.
1. LineRepository에서 너무 많은 DAO를 알고 있다.
  - 현재 Line을 만들기 위해서는 모든 DAO가 필요합니다.
  - 그래서 모든 Station, Section 정보를 다 가져와서 특정 Line을 만들 때 필요한 Station과 Section을 조립해줍니다.
  - 그렇다면 Line보다 다른 객체를 더 많이 의존하고 있는 객체의 Repository에서는 모든 의존하는 테이블의 DAO를 다 알고 있어야할까? 이부분이 좀 의아합니다.
2. Line의 삭제는 비즈니스 로직일까?
  - 현재 Line을 삭제할 떄 그에 대한 Section도 삭제하도록 했습니다.
  - 이를 서비스에서 SectionRepository를 따로 호출하지 않고
  - LineRepository에서 Line을 지우기 전에 Sections 테이블에 해당 LineId의 FK를 모두 제거합니다.
  - 이게 비즈니스 로직으로 봐야될지, Repository의 영역으로 봐야될지 잘 모르겠습니다.

## 3. 📝테스트
1. 모킹 테스트
- Service와 Repository의 단위 테스트를 할때 모킹을 사용하였습니다.
- 그렇다보니, 모킹을한 객체가 어떤 행동을 할때,
  - 결과적으로 반환된 값을 확인하거나,
  - 어떤 행위(메서드가 특정인자로 실행)가 있는지 확인합니다. 
- 그렇다보니, 현재 제가 느끼기엔 어떤 테스트들은 그냥 그 내부 로직을 그대로 테스트에 옮겨놓은 것 같다는 생각이 듭니다.
- 현재 제 단위 테스트에 대한 생각을 여쭙고 싶습니다.

2.  통합(?) 테스트
- RestAssured를 사용하여 어떤 요청이 있을 때 어떤 응답이 오는지에 대한 테스트를 작성해놓았습니다.
- 하지만 생성, 수정, 삭제 같이 어떤 응답이 왔는지가 아니라 DB내부에 어떤 변화가 있는지 확인이 필요한 경우에는 어떻게 테스틀 해야 확실한 것일지에 대해 고민이 있습니다.

## 4. 🎯API 설계 (리뷰 요청 이후, 19일 오전 10시 쯤 추가한 질문)

- 이번 미션이 늦어진데에는 API 설계를 많이 수정한 게 가장 컷습니다.
- 다른 부분보다 `노선에 역을 추가한다`는 것에 대한 고민이 많았는데요.

#### 최초 PR 때 
```http
POST  /apis/v1/stations
Content-Type: application/json

{
  "up": "잠실새내",
  "down": "잠실",
  "노선": 8,
  "노선색": "초록",
  "거리": 10
}
```
- 이렇게 `어떤 노선`에 `어떤 역들`이 `어떤 구간`을 만들지 모두 보냈습니다.
  - 그래서 해당 이름들이 유니크하다는 전제하에
  - 역이 없으면 역을 만들고,
  - 노선이 없으면 노선을 만들고,
  - 그 구간이 상행 종점, 하행 종점, 중간 구간 인지를 판단해서 노선에 추가하였습니다.

#### 1차 수정 

```http
POST /stations
Content-Type: application/json

{
  "name" : "잠실",
  "position" : [
    {
      "lineId" : 1 (2호선),
      "upStationId" : 1 (잠실나루), (null 허용)
      "upStationDistance" : 5, (null 허용)
      "downStationId" : 2 (잠실새내), (null 허용)
      "dowStationnDistance" : 5 (null 허용)
    },
    {
      "lineId" : 2 (8호선),
      "upStationId" : 3 (몽촌토성), (null 허용)
      "upDistance" : 5, (null 허용)
      "downStationId" : 4 (석촌), (null 허용)
      "downDistance" : 5 (null 허용)
    }
  ]
}
```
- `하나의 역(잠실)을 등록`한다는 것에 집중했습니다.
- 추가하려는 역이 있어야할 `노선들`과 `그 옆에 있는 역들`은 이미 존재 해서 `이름보다는 Id같은 식별자`로 판단하게 했습니다.
  - 그렇다 보니 `노선 등록`이 자연스럽게 분리되었습니다.
  - 근데 추가적으로 `초기 2역 등록`하는 API를 분리해야 했습니다.
- 제가 생각하는 등록하는 행위에 제일 이상적인 느낌이였는데,
- 이렇게 하다보니까, 삭제, 수정 같은 것도 이와 비슷하게 `큰`일을 해야 자연스러웠습니다...

#### 현재
```http
POST /stations
Content-Type: application/json

{
  "name" : "잠실"
}
```

```http
POST /lines
Content-Type: application/json

{
  "lineName" : "4호선",
  "lineColor" : "파랑"
}
```
```http
POST /lines/1/stations

{
  "upStationId" : 1,
  "downStationId" : 2,
  "distance" : 5
}
```
- 최종적으로는 이렇게 API를 각각 나눠서 진행하였습니다.
- 일단 이에 대한 의견을 여쭙고 싶습니다.
- 그리고 제가 여러번 API 설계를 뒤집었는데, 그때마다 도메인 로직이 조금씩 변경되었습니다.
- 도메인은 도메인 로직대로 두고 특정 요청에 따라 그 로직을 실행하는 형식으로 진행하고 싶었습니다.
- 하지만, 제 도메인 로직은 특정 API에 대해 필요한 기능을 만들어놔서 이 API설계를 바꾸는 동안 도메인 로직이 많이 바꼈습니다 ...
- 이에 대해서 어떻게 받아 드려야할지 잘 모르겠어서 질문드립니다...

---
# 😭 다 하지 못한 부분
1. 컨트롤러의 입력값 검증
2. 예외 핸들링
3. Station 삭제 시, Section에 Fk 있으면 SQL 예외 처리
---
열심히 작성하긴 했는데, 여러 고민과 삽질을 하다보니 시간이 많이 지연되어서 제출하게 됐습니다..😭


리뷰 잘부탁드리겠습니다..🙇‍♂️🙇‍♂️🙇‍♂️
긴 글 읽어주셔서 감사합니다.

dooboocookie committed:
docs: 요구사항 수정

pjy1368 commented on README.md:
````diff
@@ -1 +1,294 @@
+}
+```
+
+### POST /stations
+- 요청
+```
+```http request 
+POST /stations
+Content-Type: application/json
+
+{
+  "name" : "잠실"
+}
+```
+- 응답
+```http request
+HTTP /1.1 201 CREATED
+```
````
시간이 되신다면 Swagger를 통해 API 문서를 자동화해 보시면 좋겠습니다!

pjy1368 commented on src/main/java/subway/domain/Distance.java:
```diff
@@ -0,0 +1,63 @@
+
+    private final int value;
+
+    private Distance() {
+        this.value = 0;
+    }
```
사용되지 않는 생성자로 보이며, 아래 검증 메소드에서 거리는 0이하면 안된다는 조건과도 상충하므로 제거하는 것이 어떨까요?

pjy1368 commented on src/main/java/subway/domain/Line.java:
```diff
@@ -1,48 +1,117 @@
+
+    public void validateNotDuplicatedStation(final Station station) {
+        List<Station> allStations = sections.findAllStationUpToDown();
+        Set<String> names = allStations.stream()
+                .map(Station::getName)
+                .collect(Collectors.toSet());
+        if (names.contains(station.getName())) {
+            throw new IllegalArgumentException(name.getValue() + "선의 중복된 역명이 존재합니다.");
+        }
```
이 로직은 Station 관련 객체에서 담당하면 어떨까요?

dooboocookie replied:
이 검증로직은 해당 `노선`에 해당 이름이 이미 있는지를 검사하는 것인데 `노선`에서 검증을하는게 조금 더 맞지 않을까요?

pjy1368 commented on src/main/java/subway/domain/LineName.java:
```diff
@@ -0,0 +1,46 @@
+        if (value == null || value.isBlank()) {
+            throw new IllegalArgumentException("역 이름은 빈 값일 수 없습니다.");
+        }
+        if (value.length() > 10) {
+            throw new IllegalArgumentException("역 이름은 10자 이하로 입력해주세요.");
+        }
```
10이라는 숫자는 상수로 빼면 어떨까요?

pjy1368 commented on src/main/java/subway/domain/Distance.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    private void validate(final int value) {
+        if (value <= 0) {
+            throw new IllegalArgumentException("거리는 0 이하일 수 없습니다.");
+        }
```
0이라는 숫자는 상수로 빼면 어떨까요?

pjy1368 commented on src/main/java/subway/domain/Station.java:
```diff
@@ -3,19 +3,27 @@
+        if (Objects.isNull(name) || name.isBlank()) {
+            throw new IllegalArgumentException("지하철 역 이름은 null 또는 공백일 수 없습니다.");
+        }
+        if (name.length() > 10) {
+            throw new IllegalArgumentException("이름은 1~10글자여야합니다.");
+        }
```
여기도 10은 상수로 빼면 좋겠습니다.

pjy1368 commented on src/main/java/subway/domain/Subway.java:
```diff
@@ -0,0 +1,71 @@
+        this.lines = lines;
+    }
+
+    // TODO: 5/17/23 처음 도메인의 아이디를 이용해본 시도
+    public Subway deleteById(final Long lineId) {
+        Line targetLine = lines.stream()
+                .filter(line -> line.getId().equals(lineId))
+                .findAny()
+                .orElseThrow(() -> new IllegalArgumentException("삭제하려는 역이 존재하지 않습니다,"));
+        List<Line> updateLines = new ArrayList<>(lines);
+        updateLines.remove(targetLine);
+        return new Subway(updateLines);
+    }
```
도메인에서는 좀더 범용적인 delete 동작을 위해서 파라미터로 Line을 받으면 어떨까요?

dooboocookie replied:
앗 이부분에 대해서 좀 궁금한 내용이 있어서 2 단계 리뷰 떄 질문 드리겠습니다 ㅠㅠㅠ

pjy1368 commented on src/main/java/subway/ui/StationController.java:
```diff
@@ -1,54 +1,59 @@
-        stationService.deleteStationById(id);
-        return ResponseEntity.noContent().build();
+    @PutMapping("/{stationId}")
+    public ResponseEntity<Void> udpateStation(@PathVariable Long stationId, @RequestBody StationRequest request) {
```
update 오타났네요 ㅎㅎ

pjy1368 commented on src/main/java/subway/repository/LineRepository.java:
```diff
@@ -0,0 +1,84 @@
+        );
+    }
+
+    public List<Line> findAll() {
+        List<LineEntity> allLines = lineDao.findAll();
+        Map<Long, Station> allStationsById = stationDao.findAll().stream()
+                .map(StationEntity::convertToStation)
+                .collect(Collectors.toMap(Station::getId, station -> station));
+        Map<Long, List<Section>> allSectionsByLindId = getAllSectionsByLineId(allStationsById);
+        return allLines.stream()
+                .map(lineEntity -> lineEntity.convertToLine(allSectionsByLindId.getOrDefault(lineEntity.getId(), new ArrayList<>())))
+                .collect(Collectors.toList());
+    }
+
+    public Line findById(final Long id) {
+        LineEntity lineEntity = lineDao.findById(id).orElseThrow(() -> new IllegalArgumentException("해당 라인이 없습니다."));
+        Map<Long, Station> allStationsById = stationDao.findAll().stream()
+                .map(StationEntity::convertToStation)
+                .collect(Collectors.toMap(Station::getId, station -> station));
+        Map<Long, List<Section>> allSectionsByLindId = getAllSectionsByLineId(allStationsById);
+        return lineEntity.convertToLine(allSectionsByLindId.getOrDefault(lineEntity.getId(), new ArrayList<>()));
+    }
```
Line 도메인이 Sections을 의존하고 있으므로 항상 조회할 때마다 연관된 Sections을 가져와 주셨군요.
만약 비즈니스 요구 사항에서 Line이 Sections을 필요로 하지 않는 경우가 많다면 Sections이 빈 리스트인 형태로 Line은 조회하고, 추후 Sections이 필요해 질때 추가로 쿼리를 날리도록 작성해 보면 어떨까요? 이를 지연 로딩이라고도 한답니다 ㅎㅎ

pjy1368 replied:
물론 지금 동작을 지우라는 의미는 아닙니당~
다만 findAll에서는 지연 로딩으로 구현하고, findAllWithSections 같은 메소드를 추가하여 즉시 로딩(연관 객체까지 모두 포함)한 형태로 구현하시면 좋겠습니다 :)
혹은 지연 로딩인 findAll은 그대로 두고 다른 곳에서 Sections을 받아와서 비즈니스 단에서 new Line(기존 Line 값, sections)을 만들거나 line에서 setSections을 사용하는 방법도 있습니다. 가장 마음에 드는 패턴을 적용해 보세요!

pjy1368 commented on src/main/java/subway/repository/SectionRepository.java:
```diff
@@ -0,0 +1,38 @@
+        this.sectionDao = sectionDao;
+    }
+
+    public void updateByLine(final Line beforeLine, final Line updateLine) {
+        List<Section> beforeSections = beforeLine.getSections();
+        beforeSections.removeAll(updateLine.getSections());
+        List<SectionEntity> beforeSectionEntities = convertToSectionEntities(beforeLine, beforeSections);
+        sectionDao.delete(beforeSectionEntities);
+
+        List<Section> updateSections = updateLine.getSections();
+        updateSections.removeAll(beforeLine.getSections());
+        List<SectionEntity> updateSectionEntities = convertToSectionEntities(updateLine, updateSections);
+        sectionDao.save(updateSectionEntities);
+    }
```
외부에서 removeAll이 적용된 Sections를 전달해주면 어떨까요? Repository에서 꼭 해야할 일은 아닐 것 같습니다~

pjy1368 commented on src/main/java/subway/application/LineService.java:
```diff
@@ -1,53 +1,53 @@
-import java.util.List;
-import java.util.stream.Collectors;
-
+@Transactional(readOnly = true)
 @Service
 public class LineService {
```
서비스 위에 트랜잭션 어노테이션을 붙이기보다는 사용하는 메소드 위에 명시적으로 적어주면 어떨까요?

pjy1368 reviewed (approved):
안녕하세요 루카!
1차 피드백 이후 엄청난 변화가 있었군요 ㅎㅎ 전체적으로 훨씬 나은 코드가 된 것 같습니다👍
개선하면 좋을 포인트들이 몇 가지 있지만, 2단계 진행하시면서 같이 반영하셔도 될 것 같아서 우선 머지하겠습니다.
PR 본문에 남겨주신 질문은 여기 답변드릴게요 :)

Q. 테이블의 ID vs 도메인의 ID vs 리소스의 ID
사용자 쪽에서 어떤 대상을 조회하기 위해 식별자가 필요합니다. 이때 명칭은 도메인과 리소스에서는 꼭 id로 가져갈 필요 없고 별도 식별자(ex. key)로 사용해도 되나 편의상 id를 리소스, 도메인, 테이블이 모두 가져가는 패턴을 주로 사용합니다. 다만 id를 DB에서 auto-increment를 많이 사용한다고 해서 id가 테이블의 것은 아니고, 식별자 값을 도메인에서 정의하여 Entity에 해당 식별자 값을 넘겨줄 수 있다는 점은 유의해 주세요!

Q. Repository의 역할과 Domain끼리의 의존
- Domain과 Entity는 서로 구분되어 있고, Domain에서는 다른 Domain을 의존할 수 있지만 Entity 다른 Entity 의존이 불가하므로 하나의 Repository에서 여러 DAO를 갖는 것은 문제가 되지 않습니다.
오히려 Line이 그 자체로 완성된 Line을 만들려면 Station과 Section 객체가 필요하므로 StationDao, SectionDao를 LineRepository에서 갖는 것은 자연스럽다고 생각합니다!
- 도메인 입장에서는 Line이 삭제되면 FK에 의해 Sections도 삭제해야하는 사실을 모릅니다. 따라서 DB 단에서 알아서 지워줄 수도 있고, 도메인 단에서 남아있는 Sections이 있으면 Line을 못 지운다고 예외를 던지는 방법도 있습니다. 전자는 구현하기 편하지만 의도치 않는 삭제 동작이 발생할 수 있고, 후자는 API 호출 횟수가 많아질 수 있다는 단점이 있습니다. 적절히 상황에 따라 판단하여 적용하시면 됩니다.

Q. 테스트
- 원래 처음 모킹하면 그러한 느낌이 드는 것이 정상입니다 ㅎㅎ 저는 올바르게 작성하였다고 생각이 들고, 우선은 모킹을 통한 서비스 단위 테스트와 모킹 하지 않고 서비스 통합 테스트 둘 다 작성해보시는 연습을 하면 느끼는 것이 있으실 것 같습니다.
- DB 내부에 어떤 변화가 있는지는 Repository Test에서 보아야할 일이 아닐까요? RestAssured를 통한 E2E 테스트에서는 API 응답이 의도한 대로 날아왔는지 확인하시면 됩니다.

Q. API 설계
가능한 적은 일을 하도록 API 설계를 잘 하신 것 같습니다! 또한, API 동작이 도메인의 행위와 밀접한 연관이 있으므로 어느 정도 수정되는 것은 맞습니다. 특히 Service는 많이 바뀌는 것이 맞고, 도메인은 가능한 적게 바뀌도록 작성하는 것이 좋지만 API 설계 경험이 적으시다보니 이번에 겪으면서 어떻게하면 도메인 레이어는 수정을 적게 할 수 있을지 고민해보시면 좋을 것 같습니다~

추가) 다하지 못한 부분
2단계 미션에서 `@Value`를 통한 컨트롤러 입력값 검증, 커스텀 예외 추가 및 RestControllerAdvice 추가, SQL 예외 처리까지 같이 반영해주세요!
